File: addons/terrabrush/Scripts/Tools/LockTool.cs
#if TOOLS
using System;
using System.Collections.Generic;
using System.Linq;
using Godot;

namespace TerraBrush;

public class LockTool : ToolBase {
    public LockTool(TerraBrush terraBrush) : base(terraBrush) {}

    protected override ImageTexture GetToolCurrentImageTexture(ZoneResource zone) {
        if (zone.LockTexture == null) {
            zone.LockTexture = ZoneUtils.CreateLockImage(_terraBrush.ZonesSize, zone.ZonePosition);
        }

        return zone.LockTexture;
    }

    public override void Paint(TerrainToolType toolType, Image brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
        ForEachBrushPixel(brushImage, brushSize, imagePosition, (imageZoneInfo, pixelBrushStrength) => {
            if (pixelBrushStrength > 0.0) {
                var currentPixel = imageZoneInfo.Image.GetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y);
                var lockValue = toolType == TerrainToolType.LockAdd ? 1f : 0f;

                var newValue = new Color(
                    Mathf.Lerp(currentPixel.R, lockValue, pixelBrushStrength * brushStrength),
                    currentPixel.G,
                    currentPixel.B,
                    currentPixel.A
                );

                imageZoneInfo.Image.SetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y, newValue);
            }
        }, true);

        _terraBrush.TerrainZones.UpdateLockTexture();
    }
}
#endif

File: src/editor_tools/lock_tool.cpp
#include "lock_tool.h"
#include "../misc/zone_utils.h"

using namespace godot;

void LockTool::_bind_methods() {}

LockTool::LockTool() {}

LockTool::~LockTool() {}

Ref<Image> LockTool::getToolCurrentImage(Ref<ZoneResource> zone) {
    if (zone->get_lockTexture().is_null()) {
        zone->set_lockTexture(ZoneUtils::createLockImage(_terraBrush->get_zonesSize(), zone->get_zonePosition()));
    }

    return zone->get_lockTexture();
}

void LockTool::paint(TerrainToolType toolType, Ref<Image> brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
    forEachBrushPixel(brushImage, brushSize, imagePosition, ([&](ImageZoneInfo &imageZoneInfo, float pixelBrushStrength) {
        if (pixelBrushStrength > 0.0) {
            Color currentPixel = imageZoneInfo.image->get_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y);
            float lockValue = toolType == TerrainToolType::TERRAINTOOLTYPE_LOCKADD ? 1 : 0;

            Color newValue = Color(
                Math::lerp(currentPixel.r, lockValue, pixelBrushStrength * brushStrength),
                currentPixel.g,
                currentPixel.b,
                currentPixel.a
            );

            imageZoneInfo.image->set_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y, newValue);
        }
    }), true);

    _terraBrush->get_terrainZones()->updateLockTexture(_terraBrush->get_zonesSize());
}
